File: product-of-array-except-self.rb
# @param {Integer[]} nums
# @return {Integer[]}
def product_except_self(nums)
  size = nums.size

  products = [1]
  (size - 1).times do |i|
    products << products.last * nums[size - 1 - i]
  end
  products.reverse!

  left_num = 1
  size.times do |i|
    products[i] *= left_num
    left_num *= nums[i]
  end

  products
end

File: product-of-array-except-self.cc
class Solution {
public:
  vector<int> productExceptSelf(vector<int>& nums) {
    int N = nums.size();
    vector<int> ans(N, 1);
    for (int i = 1; i < N; ++i) ans[i] = ans[i - 1] * nums[i - 1];

    int right = 1;
    for (int i = N - 1; i >= 0; --i) {
      ans[i] *= right;
      right *= nums[i];
    }
    return ans;
  }
};
